src/lib/comms/include/sol-mqtt.h

```diff
@@ -0,0 +1,393 @@
+ *
+ * @{
+ */
+
+/**
+ * @struct sol_mqtt
+ *
+ * @brief MQTT Object
+ *
+ * @see sol_mqtt_connect
+ *
+ * This object is the abstraction of a MQTT session. This is the base
+ * structure for all MQTT operations and is obtained through the
+ * sol_mqtt_connect API.
+ */
+struct sol_mqtt;
+
+/**
+ * @struct sol_mqtt_message
+ *
+ * @brief MQTT Message
+ *
+ * This object is the abstraction of a MQTT message and is the base for
+ * publishing and receiving data to/from the broker.
+ */
+struct sol_mqtt_message;
+
+/**
+ * @typedef sol_mqtt_qos
+ *
+ * @brief MQTT QOS level for message delivery
+ */
+typedef enum {
+    /**
+     * The message is delivered according to the capabilities of the
+     * underlying network. No response is sent by the receiver and no
+     * retry is performed by the sender. The message arrives at the
+     * receiver either once or not at all.
+     */
+    SOL_MQTT_QOS_AT_MOST_ONCE = 0,
+
+    /**
+     * This quality of service ensures that the message arrives at the
+     * receiver at least once. A QoS 1 PUBLISH Packet has a Packet
+     * Identifier in its variable header and is acknowledged by a PUBACK
+     * Packet.
+     */
+    SOL_MQTT_QOS_AT_LEAST_ONCE = 1,
+
+    /**
+     * This is the highest quality of service, for use when neither
+     * loss nor duplication of messages are acceptable. There is an
+     * increased overhead associated with this quality of service.
+     */
+    SOL_MQTT_QOS_EXACTLY_ONCE = 2,
+} sol_mqtt_qos;
+
+/**
+ * @enum sol_mqtt_conn_status
+ *
+ * @brief Connection status
+ */
+enum sol_mqtt_conn_status {
+    /**
+     * Disconnected due to unexpected reasons.
+     */
+    SOL_MQTT_DISCONNECTED = -1,
+
+    /**
+     * Succesfully conencted to the broker
+     */
+    SOL_MQTT_CONNECTED = 0,
+
+    /**
+     * MQTT protocol rejected by the broker
+     */
+    SOL_MQTT_WRONG_PROTOCOL = 1,
+
+    /**
+     * Client ID rejected by the broker
+     */
+    SOL_MQTT_ID_REJECTED = 2,
+
+    /**
+     * Broker anavailable at provided host
+     */
+    SOL_MQTT_UNAVAILABLE = 3,
+};
+
+/**
+ * @struct sol_mqtt_config
+ *
+ * @brief Server Configuration and callback handlers
+ */
+struct sol_mqtt_config {
+#define SOL_MQTT_CONFIG_API_VERSION (1)
+    /**
+     * Should always be set to SOL_MQTT_CONFIG_API_VERSION
+     */
+    uint16_t api_version;
+
+    /**
+     * If set, the broker will drop all messages and subscriptions when
+     * the client disconnects. Must be set if no client id is provided.
+     */
+    bool clean_session;
+
+    /**
+     * Time interval between PING messages that should be sent by the
+     * broker to the client.
+     */
+    time_t keepalive;
+
+    /**
+     * NULL terminated string that should be used as client ID. If
+     * not set, clean_session must be set to true.
+     */
+    char *client_id;
+
+    /**
+     * A message that the broker should send when the client disconnects.
+     */
+    struct sol_mqtt_message *will;
```

Comment:
It's onyl needed during connect. mosquitto will keep an internal copy of the relevant data.